demos/custom-tabs-example-app/src/main/java/org/chromium/customtabsdemos/EngagementSignalsActivity.java

```diff
@@ -0,0 +1,193 @@
+            mTextVerticalScroll.setText("vertical scroll " + (isDirectionUp ? "UP️" : "DOWN"));
+        }
+
+        @Override
+        public void onGreatestScrollPercentageIncreased(int scrollPercentage, @NonNull Bundle extras) {
+            Log.d(TAG, "scroll percentage: " + scrollPercentage + "%");
+            mTextGreatestPercentage.setText("scroll percentage: " + scrollPercentage + "%");
+        }
+
+        @Override
+        public void onSessionEnded(boolean didUserInteract, @NonNull Bundle extras) {
+            Log.d(TAG, "onSessionEnded (didUserInteract=" + didUserInteract + ')');
+            mTextSessionEnd.setText(didUserInteract ? "session ended with user interaction" : "session ended without user interaction");
+        }
+    };
+
+    private CustomTabsCallback mCustomTabsCallback = new CustomTabsCallback() {
+        @Override
+        public void onNavigationEvent(int navigationEvent, @Nullable Bundle extras) {
+            String event;
+            switch (navigationEvent) {
+                case CustomTabsCallback.NAVIGATION_ABORTED:
+                    event = "NAVIGATION_ABORTED";
+                    break;
+                case CustomTabsCallback.NAVIGATION_FAILED:
+                    event = "NAVIGATION_FAILED";
+                    break;
+                case CustomTabsCallback.NAVIGATION_FINISHED:
+                    event = "NAVIGATION_FINISHED";
+                    break;
+                case CustomTabsCallback.NAVIGATION_STARTED:
+                    event = "NAVIGATION_STARTED";
+                    break;
+                case CustomTabsCallback.TAB_SHOWN:
+                    event = "TAB_SHOWN";
+                    break;
+                case CustomTabsCallback.TAB_HIDDEN:
+                    event = "TAB_HIDDEN";
+                    break;
+                default:
+                    event = String.valueOf(navigationEvent);
+            }
+            Log.d(TAG, "onNavigationEvent (navigationEvent=" + event + ')');
+            mTextNavigation.setText("onNavigationEvent " + event);
+        }
+    };
+
+    @Override
+    protected void onCreate(Bundle savedInstanceState) {
+        super.onCreate(savedInstanceState);
+        setContentView(R.layout.activity_engagement_signals);
+        findViewById(R.id.start_custom_tab).setOnClickListener(this);
+
+        mUrlEditText = findViewById(R.id.url);
+        mTextGreatestPercentage = findViewById(R.id.label_event_greatest_percentage);
+        mTextNavigation = findViewById(R.id.label_event_navigation);
+        mTextSessionEnd = findViewById(R.id.label_event_session_ended);
+        mTextVerticalScroll = findViewById(R.id.label_event_vertical_scroll);
+    }
+
+    @Override
+    protected void onStart() {
+        super.onStart();
+        bindCustomTabsService();
+    }
+
+    private void bindCustomTabsService() {
+        String packageName = CustomTabsHelper.getPackageNameToUse(this);
```

Comment:
It may be worth logging this package, potentially even adding a warning when connecting to a browser that doesn't support the engagement signals API.

Reply:
Added